File: src/dynamic-programming/5/solution.py
class Solution:
    # dynamic programming
    def longestPalindrome(self, s: str) -> str:
        if not s:
            return ""
        
        n = len(s)
        start, end = 0, 0
        
        def expand(l: int, r: int):
            while l >= 0 and r < n and s[l] == s[r]:
                l -= 1
                r += 1
            return l + 1, r - 1
        
        for i in range(n):
            print(i)
            l1, r1 = expand(i, i)
            l2, r2 = expand(i, i + 1)
            if r1 - l1 > end - start:
                start, end = l1, r1
            if r2 - l2 > end - start:
                start, end = l2, r2
                
        return s[start:end+1]
    
    # dynaminc programming
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if n < 2: return s
        dp = [[False] * n for _ in range(n)]
        start, maxLen = 0, 1
        for j in range(n):
            for i in range(0, j + 1):
                if s[i] == s[j] and (j - i < 2 or dp[i + 1][j - 1]):
                    dp[i][j] = True
                    if j - i + 1 > maxLen:
                        start = i
                        maxLen = j - i + 1
                        
        return s[start:start+maxLen]
        
        
        
s = Solution()
assert s.longestPalindrome("babad") == "bab"
# assert s.longestPalindrome("cbbd") == "bb"

File: src/dynamic-programming/5/solution.cpp
#include <string>
#include <utility>
#include <vector>

class Solution {
public:
	// O(n^2) time, O(1) space
	std::string longestPalindrome(std::string s) {
		const int n = s.size();
		if (n == 0) {
			return "";
		}

		int start = 0;
		int bestLen = 1;

		auto expand = [&](int l, int r) -> std::pair<int, int> {
			while (l >= 0 && r < n && s[l] == s[r]) {
				--l;
				++r;
			}
			return {l + 1, r - 1};
		};

		for (int i = 0; i < n; ++i) {
			auto [l1, r1] = expand(i, i);
			if (r1 - l1 + 1 > bestLen) {
				bestLen = r1 - l1 + 1;
				start = l1;
			}

			auto [l2, r2] = expand(i, i + 1);
			if (r2 - l2 + 1 > bestLen) {
				bestLen = r2 - l2 + 1;
				start = l2;
			}
		}

		return s.substr(start, bestLen);
	}

	// O(n^2) time, O(n^2) space
	std::string longestPalindromeDP(std::string s) {
		const int n = s.size();
		if (n == 0) {
			return "";
		}

		std::vector<std::vector<bool>> dp(n, std::vector<bool>(n, false));
		int start = 0;
		int bestLen = 1;

		for (int i = n - 1; i >= 0; --i) {
			for (int j = i; j < n; ++j) {
				if (s[i] == s[j] && (j - i < 3 || dp[i + 1][j - 1])) {
					dp[i][j] = true;
					if (j - i + 1 > bestLen) {
						bestLen = j - i + 1;
						start = i;
					}
				}
			}
		}

		return s.substr(start, bestLen);
	}

	// O(n^3) time, O(n) space
	std::string longestPalindromeBrute(std::string s) {
		std::string res;
		for (int i = 0; i < (int)s.size(); ++i) {
			for (int j = i; j < (int)s.size(); ++j) {
				std::string tmp = s.substr(i, j - i + 1);
				if (isPalindrome(tmp) && tmp.size() > res.size()) {
					res = tmp;
				}
			}
		}
		return res;
	}

private:
	bool isPalindrome(const std::string& s) {
		int l = 0;
		int r = s.size() - 1;
		while (l < r) {
			if (s[l] != s[r]) {
				return false;
			}
			++l;
			--r;
		}
		return true;
	}
};
